[Merged by Bors] - feat(linear_algebra): `det (M ⬝ N ⬝ M') = det N`, where `M'` is an inverse of `M`

Vierkantor opened the pull request:
This is an important step towards showing the determinant of `linear_map.to_matrix` does not depend on the choice of basis.
    
The main difficulty is allowing the two indexing types of `M` to be (a priori) different. They are in bijection though (using `basis.index_equiv` from #7631), so using `reindex_linear_equiv` we can turn everything into square matrices and apply the "usual" proof.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #7631
- [x] depends on: #7685

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
Auxiliary lemmas

Vierkantor committed:
feat(linear_algebra): det (M ⬝ N ⬝ M') = det N, where `M'` is an inverse of `M`

This is an important step towards showing the determinant of `linear_map.to_matrix` does not depend on the choice of basis.

The main difficulty is allowing the two indexing types of `M` to be (a priori) different. They are in bijection though (using `basis.index_equiv` from #7631), so using `reindex_linear_equiv` we can turn everything into square matrices and apply the "usual" proof.

Vierkantor committed:
Lint fix

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -52,6 +54,56 @@ variables {M' : Type*} [add_comm_group M'] [module R M']
+  left_inv := λ x, by rw [← matrix.to_lin'_mul_apply, hMM', matrix.to_lin'_one, id_apply],
+  right_inv := λ x, by rw [← matrix.to_lin'_mul_apply, hM'M, matrix.to_lin'_one, id_apply],
+  .. matrix.to_lin' M' } in
+equiv_of_pi_lequiv_pi e
```
This looks like it's maybe two definitions to me:
```suggestion
/-- If `M` and `M'` are each other's inverse matrices, they provide an equivalence between `m → A`
and `n → A` corresponding to `M.mul_vec` and `M'.mul_vec`. -/
def matrix.to_lin'_of_inv [decidable_eq m] [decidable_eq n]
  {M : matrix m n A} {M' : matrix n m A}
  (hMM' : M ⬝ M' = 1) (hM'M : M' ⬝ M = 1) :
  (m → A) ≃ₗ[A] (n → A) :=
{ to_fun := matrix.to_lin' M',
  inv_fun := M.to_lin',
  left_inv := λ x, by rw [← matrix.to_lin'_mul_apply, hMM', matrix.to_lin'_one, id_apply],
  right_inv := λ x, by rw [← matrix.to_lin'_mul_apply, hM'M, matrix.to_lin'_one, id_apply],
  .. matrix.to_lin' M' }

/-- If `M` and `M'` are each other's inverse matrices, they are square matrices up to
equivalence of types. -/
def matrix.index_equiv_of_inv [decidable_eq m] [decidable_eq n]
  {M : matrix m n A} {M' : matrix n m A}
  (hMM' : M ⬝ M' = 1) (hM'M : M' ⬝ M = 1) :
  m ≃ n :=
equiv_of_pi_lequiv_pi (matrix.to_lin'_of_inv hMM' hM'M)
```
I think the first definition belongs in `linear_algebra/matrix/to_lin.lean`?

Vierkantor committed:
Add `matrix.to_lin'_of_inv` and `matrix.to_lin_of_inv`

If two a priori rectangular matrices are each other's inverses, the corresponding linear map is in fact a linear equiv.

Co-Authored-By: Eric Wieser <wieser.eric@gmail.com>

Vierkantor committed:
Use `to_lin'_of_inv` in `det_conjugate`

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -52,6 +54,50 @@ variables {M' : Type*} [add_comm_group M'] [module R M']
+      matrix.reindex_linear_equiv_mul e (equiv.refl n) e,
+      matrix.reindex_linear_equiv_mul e (equiv.refl n) (equiv.refl n),
+      matrix.reindex_linear_equiv_refl_refl, linear_equiv.refl_apply]
+end
```
Does this belong in `linear_algebra/matrix/determinant` instead?

Vierkantor replied:
This is the first place in the import hierarchy where it fits: `matrix/to_lin.lean` and `matrix/determinant.lean` do not import each other.

You can also see this as an auxiliary lemma for the definition of `linear_map.det` in #7635, which definitely fits in this file.

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -52,6 +54,50 @@ variables {M' : Type*} [add_comm_group M'] [module R M']
+      matrix.reindex_linear_equiv_mul e (equiv.refl n) e,
+      matrix.reindex_linear_equiv_mul e (equiv.refl n) (equiv.refl n),
+      matrix.reindex_linear_equiv_refl_refl, linear_equiv.refl_apply]
+end
```
I was able to show this using the lemmas in #7685 with:
```suggestion
begin
  -- Although `m` and `n` are different a priori, we will show they have the same cardinality.
  -- This turns the problem into one for square matrices (`matrix.det_units_conj`), which is easy.
  let e : m ≃ n := matrix.index_equiv_of_inv hMM' hM'M,
  let U : units (matrix n n A) := ⟨
    M.minor e.symm (equiv.refl _),
    M'.minor (equiv.refl _) e.symm,
    by rw [mul_eq_mul, ←minor_mul_equiv, hMM', minor_one_equiv],
    by rw [mul_eq_mul, ←minor_mul_equiv, hM'M, minor_one_equiv]⟩,
  rw [
    ←matrix.det_units_conj U N,
    ←det_minor_equiv_self e.symm,
    minor_mul_equiv _ _ _ (equiv.refl _) _,
    minor_mul_equiv _ _ _ (equiv.refl _) _],
  refl,
end
```
Do you think this is better or worse than your proof?

Vierkantor replied:
I would like to see one proved in terms of the other, although it doesn't matter too much which one is first. The way I see it, either we push all the complexity to the "hard" `det_conjugate` case and let `det_units_conj` be a direct corollary, or we can do the "easy" `det_units_consj` case first, then show in `det_conjugate` that the hard case reduces to the easy case. Neither is obviously better than the other to me.

Vierkantor replied:
In the interest of getting things in the PR queue moving again, I'm merging your PR and rebasing mine on top of it.

Vierkantor committed:
Lint fix

Vierkantor committed:
Use `det_units_conj` to prove `det_conjugate`

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -52,6 +54,47 @@ variables {M' : Type*} [add_comm_group M'] [module R M']
+equiv_of_pi_lequiv_pi (matrix.to_lin'_of_inv hMM' hM'M)
+
+/-- If `M'` is a two-sided inverse for `M` (indexed differently), `det (M ⬝ N ⬝ M') = det N`. -/
+lemma matrix.det_conjugate
```
Thoughts on `conj` vs `conjugate`? I think the former is more common (eg mul_aut.conj)

Vierkantor replied:
Looks like the convention (more consistent than I recall!) is that `conj` is for "a * b * a⁻¹" and `conjugate` is for "conjugate roots", so I'll change it.

eric-wieser commented on src/linear_algebra/matrix/to_lin.lean:
```diff
@@ -135,6 +135,11 @@ by { ext, rw [matrix.one_apply, linear_map.to_matrix'_apply, id_apply] }
   matrix.to_lin' (M ⬝ N) = (matrix.to_lin' M).comp (matrix.to_lin' N) :=
 by { ext, simp }
 
+/-- Shortcut lemma for `matrix.to_lin'_mul_apply` and `linear_map.comp_apply` -/
```
I think this comment is wrong, it says it is a shortcut for itself!

eric-wieser replied:
```suggestion
/-- Shortcut lemma for `matrix.to_lin'_mul` and `linear_map.comp_apply` -/
```

eric-wieser reviewed (approved):
bors d+

Thanks!

Vierkantor committed:
Fix docstring

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Vierkantor committed:
Rename `det_conjugate` -> `det_conj`
